src/librustc_infer/infer/mod.rs

```diff
@@ -636,18 +700,22 @@ impl<'tcx> InferOk<'tcx, ()> {
     }
 }
 
+/// Extends `CombinedSnapshot` by tracking which variables were added in the snapshot
 #[must_use = "once you start a snapshot, you should always consume it"]
-pub struct CombinedSnapshot<'a, 'tcx> {
-    projection_cache_snapshot: traits::ProjectionCacheSnapshot,
-    type_snapshot: type_variable::Snapshot<'tcx>,
-    const_snapshot: ut::Snapshot<ut::InPlace<ty::ConstVid<'tcx>>>,
-    int_snapshot: ut::Snapshot<ut::InPlace<ty::IntVid>>,
-    float_snapshot: ut::Snapshot<ut::InPlace<ty::FloatVid>>,
+pub struct FudgeSnapshot<'a, 'tcx> {
+    snapshot: CombinedSnapshot<'a, 'tcx>,
     region_constraints_snapshot: RegionSnapshot,
-    region_obligations_snapshot: usize,
+    type_snapshot: type_variable::Snapshot<'tcx>,
```

Comment:
So, this type sounds like it encodes a lot of information, but really it just encodes a length, and the same is true for `RegionSnapshot` -- so essentially this `FudgeSnapshot` is just adding in the length information? 

Is there any data to suggest that storing 5 or 6 integers meaningfully changes performance? If not, it seems to me like it's not worth adding the `probe` vs `probe_fudge` distinction.

Reply:
> So, this type sounds like it encodes a lot of information, but really it just encodes a length, and the same is true for RegionSnapshot -- so essentially this FudgeSnapshot is just adding in the length information?

Yes, should perhaps refactor that a bit to make it clear it is just the length.

> Is there any data to suggest that storing 5 or 6 integers meaningfully changes performance? If not, it seems to me like it's not worth adding the probe vs probe_fudge distinction.

I don't have any specific data, and I doubt it makes a large difference. However each individual snapshot that was captured before is just capturing two integers and incrementing a third so these 5-6 integers is roughly as expensive as two snapshots. Since there is evidently gain in reducing how many snapshots are done, doing this relatively small change to reduce them a bit further seemed worthwhile to me (snapshots are done a lot!).

`probe_fudge` is actually only called in the entire compiler so it is definitely a special case, arguably in simplifies things for all the other `probe` uses which do not care about this.

Reply:
Well, it doesn't really change anything for the callers, right, since they can't even access those fields in any particular way? Indeed, `probe_fudge` could be private to the `infer` module, I think?

I guess the main change is that it adds this extra concept, a `FudgeSnapshot`, and another method, `probe_fudge`, which then means that users have to think about whether that is something that they might want. Privacy helps here some, though not that much.

It's true that snapshots occur a lot, but I assume that their associated costs don't come so much from the actual number of bytes in the snapshot structure.

Reply:
> Well, it doesn't really change anything for the callers, right, since they can't even access those fields in any particular way? Indeed, probe_fudge could be private to the infer module, I think?

True, moved it into the fudge module. Also extracted these lengths fields into a struct separate from snapshotting since they do not not truly have anything to do with it.

Reply:
OK, yes, I like this reasonably well. Thanks @Marwes .